includes/class-wc-stripe-payment-tokens.php

```diff
@@ -160,6 +209,122 @@ public function woocommerce_get_customer_payment_tokens( $tokens, $customer_id,
 		return $tokens;
 	}
 
+	/**
+	 * Gets saved tokens from Intentions API if they don't already exist in WooCommerce.
+	 *
+	 * @param array  $tokens     Array of tokens
+	 * @param string $user_id    WC User ID
+	 * @param string $gateway_id WC Gateway ID
+	 *
+	 * @return array
+	 */
+	public function woocommerce_get_customer_upe_payment_tokens( $tokens, $user_id, $gateway_id ) {
+		if ( is_user_logged_in() ) {
+			if ( count( $tokens ) >= get_option( 'posts_per_page' ) ) {
+				// The tokens data store is not paginated and only the first "post_per_page" (defaults to 10) tokens are retrieved.
+				// Having 10 saved credit cards is considered an unsupported edge case, new ones that have been stored in Stripe won't be added.
+				return $tokens;
+			}
+
+			$gateway                  = new WC_Stripe_UPE_Payment_Gateway();
+			$reusable_payment_methods = array_filter( $gateway->get_upe_enabled_payment_method_ids(), [ $gateway, 'is_enabled_for_saved_payments' ] );
+			$customer                 = new WC_Stripe_Customer( $user_id );
+			$remaining_tokens         = [];
+
+			foreach ( $tokens as $token ) {
+				$payment_method_type = $this->get_payment_method_type_from_token( $token );
+				if ( ! in_array( $payment_method_type, $reusable_payment_methods, true ) ) {
+					// Remove saved token from list, if payment method is not enabled.
+					unset( $tokens[ $token->get_id() ] );
+				} else {
+					// Store relevant existing tokens here.
+					// We will use this list to check whether these methods still exist on Stripe's side.
+					$remaining_tokens[ $token->get_token() ] = $token;
+				}
+			}
+
+			$retrievable_payment_method_types = [];
+			foreach ( $reusable_payment_methods as $payment_method_id ) {
+				$upe_payment_method = $gateway->payment_methods[ $payment_method_id ];
+				if ( ! in_array( $upe_payment_method->get_retrievable_type(), $retrievable_payment_method_types, true ) ) {
+					$retrievable_payment_method_types[] = $upe_payment_method->get_retrievable_type();
```

Comment:
We will only ever need to query Stripe for payment methods that are `card` or `sepa_debit`, though some of the latter may have been generated from other payment methods.